File: examples/source/atomic/atomic.rs
// use std::sync::atomic::{AtomicI32, Ordering};

// pub fn fetch_add(counter: &AtomicI32, value: i32) -> i32 {
//     counter.fetch_add(value, Ordering::Relaxed)
// }

// pub fn load(counter: &AtomicI32) -> i32 {
//     counter.load(Ordering::Relaxed)
// }

// pub fn store(counter: &AtomicI32, value: i32) {
//     counter.store(value, Ordering::Relaxed)
// }

// pub fn fetch_add(counter: &mut i32, value: i32) -> i32 {
//     let old = unsafe { std::ptr::read_volatile(counter) };
//     unsafe { std::ptr::write_volatile(counter, old + value) };
//     old
// }

// pub fn load(counter: &i32) -> i32 {
//     unsafe { std::ptr::read_volatile(counter) }
// }

// pub fn store(counter: &mut i32, value: i32) {
//     unsafe { std::ptr::write_volatile(counter, value) }
// }

pub fn fetch_add(counter: &mut i32, value: i32) -> i32 {
    let old = *counter;
    *counter = old.wrapping_add(value);
    old
}

pub fn load(counter: &i32) -> i32 {
    *counter
}

pub fn store(counter: &mut i32, value: i32) {
    *counter = value;
}

File: examples/source/atomic/atomic.cpp
//! unfortunately, Alive2 does NOT support atomic operations and
//! the corresponding LLVM IR instructions.

#include <atomic>

// int fetch_add(std::atomic<int>& counter, int value) {
//     return counter.fetch_add(value, std::memory_order_relaxed);
// }

// int load(std::atomic<int>& counter) {
//     return counter.load(std::memory_order_relaxed);
// }

// void store(std::atomic<int>& counter, int value) {
//     counter.store(value, std::memory_order_relaxed);
// }

// use volatile to prevent optimization
// int fetch_add(volatile int* counter, int value) {
//     int old = *counter;
//     *counter = old + value;
//     return old;
// }

// int load(volatile int* counter) {
//     return *counter;
// }

// void store(volatile int* counter, int value) {
//     *counter = value;
// }

int fetch_add(int* counter, int value) {
    int old = *counter;
    *counter = old + value;
    return old;
}

int load(const int* counter) {
    return *counter;
}

void store(int* counter, int value) {
    *counter = value;
}
